internal/display/bufwindow.go

```diff
@@ -106,51 +106,35 @@ func (w *BufWindow) Clear() {
 	}
 }
 
-// Bottomline returns the line number of the lowest line in the view
-// You might think that this is obviously just v.StartLine + v.Height
-// but if softwrap is enabled things get complicated since one buffer
-// line can take up multiple lines in the view
-func (w *BufWindow) Bottomline() int {
-	if !w.Buf.Settings["softwrap"].(bool) {
-		h := w.StartLine + w.Height - 1
-		if w.drawStatus {
-			h--
-		}
-		return h
-	}
-
-	l := w.LocFromVisual(buffer.Loc{0, w.Y + w.Height})
-
-	return l.Y
-}
-
 // Relocate moves the view window so that the cursor is in view
 // This is useful if the user has scrolled far away, and then starts typing
 // Returns true if the window location is moved
 func (w *BufWindow) Relocate() bool {
 	b := w.Buf
-	// how many buffer lines are in the view
-	height := w.Bottomline() + 1 - w.StartLine
-	h := w.Height
+	height := w.Height
 	if w.drawStatus {
-		h--
+		height--
 	}
 	ret := false
 	activeC := w.Buf.GetActiveCursor()
-	cy := activeC.Y
 	scrollmargin := int(b.Settings["scrollmargin"].(float64))
-	if cy < w.StartLine+scrollmargin && cy > scrollmargin-1 {
-		w.StartLine = cy - scrollmargin
+
+	c := w.SLocFromLoc(activeC.Loc)
+	bStart := SLoc{0, 0}
+	bEnd := w.SLocFromLoc(b.End())
+
+	if c.LessThan(w.Scroll(w.StartLine, scrollmargin)) && c.GreaterThan(w.Scroll(bStart, scrollmargin-1)) {
+		w.StartLine = w.Scroll(c, -scrollmargin)
 		ret = true
-	} else if cy < w.StartLine {
-		w.StartLine = cy
+	} else if c.LessThan(w.StartLine) {
+		w.StartLine = c
 		ret = true
 	}
-	if cy > w.StartLine+height-1-scrollmargin && cy < b.LinesNum()-scrollmargin {
-		w.StartLine = cy - height + 1 + scrollmargin
+	if c.GreaterThan(w.Scroll(w.StartLine, height-1-scrollmargin)) && c.LessThan(w.Scroll(bEnd, -scrollmargin+1)) {
+		w.StartLine = w.Scroll(c, -height+1+scrollmargin)
 		ret = true
-	} else if cy >= b.LinesNum()-scrollmargin && cy >= height {
-		w.StartLine = b.LinesNum() - height
+	} else if c.GreaterThan(w.Scroll(bEnd, -scrollmargin)) && c.GreaterThan(w.Scroll(w.StartLine, height-1)) {
```

Comment:
Note that I've changed `cy >= height` to essentially `cy >= w.StartLine + height`. This, together with the removal of `Bottomline`, fixes some bugs with scrolling past the last line.